otel-agent/ecs-ec2/ecslogresourcedetectionprocessor/ecslog_resourcedetection_processor.go

```diff
@@ -0,0 +1,121 @@
+package ecslogresourcedetectionprocessor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/coralogix/telemetry-shippers/otel-agent/ecs-ec2/ecslogresourcedetectionprocessor/internal/metadata"
+	"github.com/coralogix/telemetry-shippers/otel-agent/ecs-ec2/ecslogresourcedetectionprocessor/internal/utils"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/processor/processorhelper"
+	"go.uber.org/zap"
+)
+
+// global variable used to store metadata endpoints within package
+var ecsMetadataHandler metadataHandler
+
+func processLogsFunc(logger *zap.Logger, c *Config) processorhelper.ProcessLogsFunc {
+	return func(ctx context.Context, ld plog.Logs) (plog.Logs, error) {
+		for i := 0; i < ld.ResourceLogs().Len(); i++ {
+			rlog := ld.ResourceLogs().At(i)
+			containerID := getContainerId(&rlog)
+			logger.Debug("processing",
+				zap.String("container.id", containerID),
+				zap.String("processor", metadata.Type))
+
+			// check for ecs agent, Note that ecs agents do not have
+			// metadata endpoints
+			if ok, _ := isECSAgent(containerID); ok {
+				rlog.Resource().Attributes().PutStr("ecs.agent", "true")
+			}
+
+			endpoints, ok := ecsMetadataHandler.get(containerID)
+			if !ok || len(endpoints) == 0 {
+				logger.Debug("metadata not found",
+					zap.String("container.id", containerID),
+					zap.String("processor", metadata.Type))
+				return ld, nil
+			}
+
+			resp, err := http.Get(endpoints[0])
```

Comment:
Thanks @povilasv 

Good point, technically, the metadata won't change until the container is destroyed which means we can store the metadata directly. The endpoints are synchonized every 60s, so we can do a set of API calls to collect all the metadata then.

Will look at implementing this.

To my knowledge, there's no cost to calling the EC2/ECS metadata endpoints.

Reply:
@povilasv, 
I've implemented caching. 
- When the processor starts it will record all metadata for existing containers. T
- The processor will also monitor the docker.sock and listen for container create events, on a new create event it will re-sync the metadata.
- The processor also re-syncs every 60s.